text/0000-read-all.md

```diff
@@ -0,0 +1,99 @@
+- Feature Name: read_exact and read_full
+- Start Date: 2015-03-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Rust's `Write` trait has `write_all`, which is a convenience method that calls
+`write` repeatedly to write an entire buffer. This proposal adds two similar
+convenience methods to the `Read` trait: `read_full` and `read_exact`.
+`read_full` calls `read` repeatedly until the buffer has been filled, EOF has
+been reached, or an error other than `Interrupted` occurs. `read_exact` is
+similar to `read_full`, except that reaching EOF before filling the buffer is
+considered an error.
+
+# Motivation
+
+The `read` method may return fewer bytes than requested, and may fail with an
+`Interrupted` error if a signal is received during the call. This requires
+programs wishing to fill a buffer to call `read` repeatedly in a loop. This is
+a very common need, and it would be nice if this functionality were provided in
+the standard library. Many C and Rust programs have the same need, and solve it
+in the same way. For example, Git has [`read_in_full`][git], which behaves like
+the proposed `read_full`, and the Rust byteorder crate has
+[`read_full`][byteorder], which behaves like the proposed `read_exact`.
+[git]: https://github.com/git/git/blob/16da57c7c6c1fe92b32645202dd19657a89dd67d/wrapper.c#L246
+[byteorder]: https://github.com/BurntSushi/byteorder/blob/2358ace61332e59f596c9006e1344c97295fdf72/src/new.rs#L184
+
+# Detailed design
+
+The following methods will be added to the `Read` trait:
+
+``` rust
+fn read_full(&mut self, buf: &mut [u8]) -> Result<usize>;
```

Comment:
@alexcrichton 

> It's not composable with all existing uses of `T: Read`, everyone now needs to remember to call `read_full` instead of `read`.

Can you explain this a bit more? Why isn't it composable? Why would anyone need remember to call `read_full` if they didn't want to? Like `read_exact`, it's a just convenience method for filling a buffer, implemented by calling `read`.

> It's specific to just readers when writers also likely want the ability to "just ignore EINTR"

First, it's not "just ignore EINTR". It's looping until the entire buffer has been filled (unless an EOF occurrs). `read` is not guaranteed to return as many bytes as you ask for, even if you don't encounter an EOF. For example, if you ask `read` for 10 bytes, but only 5 are currently available, the implementation is free to just return those 5 bytes and put off blocking until the next call. Being guaranteed (unless an error occurs) that the only time your buffer won't be filled is an EOF is very useful.

Second, `Write` already has `write_all` which covers this use case. Reading and writing are different. When writing, the program determines what bytes need to be written, and it will almost always be an error when it can't write all of them. When reading, it is the file or remote sender determining the bytes to be sent, and not having enough bytes before EOF to fill the buffer may or may not be an error depending on the file format or protocol. That is why it makes sense for `Write` to have `write_all` and `Read` to have

> This is a lossy method because if an error is encountered you have no way of knowing how much data was previously written.

Yes. This is well understand by the discussion so far and (as far as I can tell) generally considered not to be a problem. These are convenience methods, after all, and the vast majority of the time one will just want to fail if an IO error occurs. If you need more information, call `read` directly.

> If we were to add a function like this I believe we would add something akin to a `uninterrupted()` adapter.

This doesn't solve the problem. Code would still have to loop to make sure the buffer was filled, and the programmer would have to know the type of the reader to know if they can even avoid checking for `Interrupted`, and what's the point if they have to loop anyway? `read_full`, on the other hand, can always be relied upon to work for any reader, with no extra effort for the implementer.

In summary, both `read_full` and `read_exact` are needed as `Read` equivalents of `write_all` due to the fundamental differences between reading and writing. The cases that distinguish them don't exist when writing.